trunk/source/WitnessPrinter/src/de/uni_freiburg/informatik/ultimate/witnessprinter/yaml/YamlCorrectnessWitnessGenerator.java

```diff
@@ -40,72 +34,62 @@ public class YamlCorrectnessWitnessGenerator {
+	private final FormatVersion mFormatVersion;
+	private final boolean mIsAcslAllowed;
+	private final Map<String, String> mProgramHashes;
+	private final YamlWitnessWriter mWriter;
 
 	public YamlCorrectnessWitnessGenerator(final IIcfg<? extends IcfgLocation> icfg, final ILogger logger,
 			final IUltimateServiceProvider services) {
 		mLogger = logger;
 		mIcfg = icfg;
 		mPreferences = PreferenceInitializer.getPreferences(services);
-	}
-
-	private Witness getWitness() {
+		mFormatVersion =
+				FormatVersion.fromString(mPreferences.getString(PreferenceInitializer.LABEL_YAML_FORMAT_VERSION));
+		mIsAcslAllowed = mFormatVersion.compareTo(new FormatVersion(2, 1)) >= 0;
 		final String producer = mPreferences.getString(PreferenceInitializer.LABEL_GRAPH_DATA_PRODUCER);
-		final String hash = mPreferences.getString(PreferenceInitializer.LABEL_GRAPH_DATA_PROGRAMHASH);
+		final String programHash = mPreferences.getString(PreferenceInitializer.LABEL_GRAPH_DATA_PROGRAMHASH);
 		final String spec = mPreferences.getString(PreferenceInitializer.LABEL_GRAPH_DATA_SPECIFICATION);
 		final String arch = mPreferences.getString(PreferenceInitializer.LABEL_GRAPH_DATA_ARCHITECTURE);
-		final FormatVersion formatVersion =
-				FormatVersion.fromString(mPreferences.getString(PreferenceInitializer.LABEL_YAML_FORMAT_VERSION));
 		final String version = new UltimateCore().getUltimateVersionString();
 		final String filename = ILocation.getAnnotation(mIcfg).getFileName();
-		final Supplier<Metadata> metadataSupplier = () -> new Metadata(formatVersion, UUID.randomUUID(),
-				OffsetDateTime.now(), new Producer(producer, version),
-				new Task(List.of(filename), Map.of(filename, hash), spec, arch, "C"));
+		mProgramHashes = Map.of(filename, programHash);
+		mWriter = YamlWitnessWriter.construct(mFormatVersion,
+				new MetadataProvider(mFormatVersion, producer, version, mProgramHashes, spec, arch, "C"));
+	}
 
+	private Witness getWitness() {
 		final List<IcfgLocation> allProgramPoints = mIcfg.getProgramPoints().values().stream()
 				.flatMap(x -> x.values().stream()).collect(Collectors.toList());
-
-		// TODO: Should we sort these entries somehow (for consistent result in validation and to improve readability)
-		// e.g. by line number and/or entry type?
-		final List<WitnessEntry> entries = new ArrayList<>();
-		entries.addAll(extractLoopInvariants(allProgramPoints, metadataSupplier, hash, formatVersion));
-		if (formatVersion.getMajor() >= 3) {
-			entries.addAll(extractFunctionContracts(allProgramPoints, metadataSupplier, hash, formatVersion));
-		}
-		final Witness witness = new Witness(entries);
-		if (formatVersion.getMajor() < 2) {
-			return witness;
-		}
-		return new Witness(List.of(witness.toInvariantSet(metadataSupplier)));
+		final var invariants = extractInvariants(allProgramPoints);
+		final var functionContracts = extractFunctionContracts(allProgramPoints);
+		return new Witness(DataStructureUtils.concat(invariants, functionContracts));
```

Comment:
This TODO got lost during refactoring 😅
If we do not sort, we rely on the order of `mIcfg.getProgramPoints()`, which seems to be consistent for `BasicIcfg` the very least, so maybe this is not a critical issue -- it might however not hurt to keep the TODO.